syntax = "proto3";

import "google/protobuf/struct.proto";

enum Action {
    NONE = 0;
    SELECT = 1;
    CREATE = 2;
    DELETE = 3;
    UPDATE = 4;
    COUNT = 5;
}

message Query {
    Action action = 1;
    // repeated string fields = 2;
    optional google.protobuf.Struct filter_kwargs = 3;
    optional google.protobuf.Struct exclude_kwargs = 4;
    optional int64 start_index = 5;
    optional int64 end_index = 6;
    repeated string order = 7;
    optional google.protobuf.Struct object_data = 8;
    string resource_type = 9;
}

message Response {
    Action action = 1;
    google.protobuf.Struct data = 2;
}

service RemoteQueryset {
    rpc SingleQuery (Query) returns (stream Response);
    rpc TransactionalQuery (stream Query) returns (stream Response);
}